### PR TITLE
Hunspell suggestions: speed up for some non-Latin scripts

### PR DESCRIPTION
**Pull request by donnerpeter**
# Description

e.g. Russian suggestion calculation could be faster

# Solution

* don't bother with PriorityQueue management for roots that have no letters in common with the misspelled one (e.g. Latin in a Cyrilic text)
* try to tableize the trigram RunAutomaton by deducting the smallest char to hopefully make it into 0..127 range

# Tests

`TestPerformance.ru/ru_suggest` added, the latter one getting a ~30% speedup with this change on my corpus.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/lucene/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [ ] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Lucene maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.

### CODE REVIEW

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/GeneratingSuggester.java [hunk 1]
```diff
@@ -94,10 +94,12 @@ char transformChar(char c) {
             return;
           }
 
-          int sc =
-              automaton.ngramScore(rootChars)
-                  - longerWorsePenalty(word.length(), rootChars.length)
-                  + commonPrefix(word, rootChars);
+          int sc = automaton.ngramScore(rootChars);
+          if (sc == 0) {
+            return; // no common characters at all, don't suggest this root
```
Comment (donnerpeter) [inline lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/GeneratingSuggester.java:99]: It turns out that adding/removing same-zero-cost items to this PQ is quite noticeable

File: lucene/analysis/common/src/test/org/apache/lucene/analysis/hunspell/TestPerformance.java [hunk 2]
```diff
@@ -181,7 +192,8 @@ private Path findAffFile(String code) throws IOException {
         String line = reader.readLine();
         if (line == null) break;
 
-        for (String token : line.split("[^a-zA-Z" + Pattern.quote(dictionary.wordChars) + "]+")) {
+        for (String token :
+            line.split("[^\\p{IsLetter}" + Pattern.quote(dictionary.wordChars) + "]+")) {
```
Comment (donnerpeter) [inline lucene/analysis/common/src/test/org/apache/lucene/analysis/hunspell/TestPerformance.java:196]: All cyrilic letters were skipped before this change :(

File: lucene/analysis/common/src/test/org/apache/lucene/analysis/hunspell/TestPerformance.java [hunk 3]
```diff
@@ -121,6 +131,7 @@ private void checkSuggestionPerformance(String code, int wordCount) throws Excep
     Hunspell speller = new Hunspell(dictionary, TimeoutPolicy.THROW_EXCEPTION, () -> {});
     List<String> words =
         loadWords(code, wordCount, dictionary).stream()
+            .distinct()
```
Comment (donnerpeter) [inline lucene/analysis/common/src/test/org/apache/lucene/analysis/hunspell/TestPerformance.java:134]: Let's measure the performance of different word suggestions (we only check very few of them anyway, so let's ensure diversity). In spellchecking measurement OTOH, we check the text as it flows, with all the duplicates, to simulate real-life spellchecker

**Review by rmuir — APPROVED** [review]
looks good. non-latin dictionaries may get the latin fast-path (tableization) now in CharRunAutomaton.

**Review by dweiss — APPROVED** [review]

**Comment by donnerpeter** [comment]
Ouch, forgot to mention the issue LUCENE-9842 :(

**Pull request merged**